Tests/ClientRuntimeTests/NetworkingTests/ContentLengthMiddlewareTests.swift

```diff
@@ -0,0 +1,85 @@
+    }
+
+    func testContentLengthSetWhenRequiresLengthAndUnsignedPayload() async throws {
+        addContentLengthMiddlewareWith(requiresLength: true, unsignedPayload: true)
+        try await AssertHeadersArePresent(expectedHeaders: ["Content-Length": "0"])
+    }
+
+    func testRequiresLengthSetWithNilStreamShouldThrowError() async throws {
+        addContentLengthMiddlewareWith(requiresLength: true, unsignedPayload: false)
+        forceEmptyStream()
+        do {
+            try await AssertHeadersArePresent(expectedHeaders: ["Content-Length": "0"])
+            XCTFail("Should throw error")
+        } catch let error as StreamError {
+            switch error {
+            case .notSupported("sigv4 requires length"), .notSupported("operation requires length"):
+                // The error matches one of the expected cases, test passes
+                break
+            default:
+                XCTFail("Error is not StreamError.notSupported with expected message")
+            }
+        }
+    }
+
+    private func addContentLengthMiddlewareWith(requiresLength: Bool, unsignedPayload: Bool) {
+        stack.finalizeStep.intercept(
+            position: .before,
+            middleware: ContentLengthMiddleware(requiresLength: requiresLength, unsignedPayload: unsignedPayload)
+        )
+    }
+
+    private func forceEmptyStream() {
+        // Force stream length to be nil
+        stack.finalizeStep.intercept(position: .before, id: "set nil stream length") { (context, input, next) -> OperationOutput<MockOutput> in
+            input.body = .stream(BufferedStream()) // Set the stream length to nil
+            return try await next.handle(context: context, input: input)
+        }
+    }
+        
+    private func AssertHeadersArePresent(expectedHeaders: [String: String], file: StaticString = #file, line: UInt = #line) async throws -> Void {
+        _ = try await stack.handleMiddleware(context: builtContext, input: MockInput(),
+                                              next: MockHandler(handleCallback: { (_, input) in
+                                                  for (key, value) in expectedHeaders {
+                                                      XCTAssert(input.headers.value(for: key) == value, file: file, line: line)
+                                                  }
+                                                  let httpResponse = HttpResponse(body: HttpBody.none, statusCode: HttpStatusCode.ok)
+                                                  let mockOutput = try! MockOutput(httpResponse: httpResponse, decoder: nil)
+                                                  let output = OperationOutput<MockOutput>(httpResponse: httpResponse,
+                                                                                           output: mockOutput)
+                                                  return output
+                                              }))
```

Comment:
I think it would be good if `MockHandler` construction is factored outside of the `handleMiddleware` method call to make this more readable.